src/main/java/fi/okm/jod/yksilo/service/KoskiService.java

```diff
@@ -0,0 +1,138 @@
+
+@Service
+@Slf4j
+public class KoskiService {
+  public static final Set<String> ALLOWED_HOSTS =
+      Set.of("https://opintopolku.fi", "https://testiopintopolku.fi");
+
+  private final RestClient restClient;
+
+  public KoskiService(
+      RestClient.Builder restClientBuilder,
+      MappingJackson2HttpMessageConverter messageConverter,
+      @Value("") String baseUrl) {
+    log.info("Creating KoskiService, baseUrl: {}", baseUrl);
+
+    this.restClient =
+        restClientBuilder
+            .messageConverters(
+                converters -> {
+                  converters.clear();
+                  converters.add(messageConverter);
+                })
+            .baseUrl(baseUrl)
+            .defaultHeader(
+                HttpHeaders.USER_AGENT, "fi.okm.jod (https://okm.fi/hanke?tunnus=OKM069:00/2021)")
+            .build();
+  }
+
+  private <T> T readJsonProperty(Object json, String jsonPathExpression) {
+    try {
+      return JsonPath.read(json, jsonPathExpression);
+    } catch (Exception e) {
+      log.error("JSON expression error", e);
+      return null;
+    }
+  }
+
+  public List<KoulutusDto> getKoskiData(URI linkki) {
+    var h = linkki.getScheme() + "://" + linkki.getHost();
+    var location = ALLOWED_HOSTS.stream().filter(a -> a.equals(h)).findFirst();
+    if (location.isPresent()) {
+      final var koskiResponse =
+          restClient
+              .get()
+              .uri(location.get() + "/" + linkki.getPath())
+              .retrieve()
+              .body(Object.class);
+
+      List<Object> opinnot = readJsonProperty(koskiResponse, "$.opiskeluoikeudet");
+      if (opinnot == null) {
+        return Collections.emptyList();
+      }
+
+      return opinnot.stream()
+          .map(
+              o -> {
+                var toimija =
+                    readJsonProperty(o, "$.oppilaitos") != null
+                        ? readJsonProperty(o, "$.oppilaitos")
+                        : readJsonProperty(o, "$.koulutustoimija");
+                var nimet = readJsonProperty(toimija, "$.nimi");
+                var localizedNimi = getLocalizedString(nimet);
+                var suoritukset =
+                    readJsonProperty(o, "$.suoritukset[0].koulutusmoduuli.tunniste.nimi");
+                var localizedKuvaus = getLocalizedString(suoritukset);
+
+                String alkuExpression = "$.alkamispäivä";
+
+                if (linkki.getPath().contains("/suoritetut-tutkinnot/")) {
+                  // read vahvistettu
+                  alkuExpression = "$.suoritukset[0].vahvistus.päivä";
+                }
+                var start = getLocalDate(o, alkuExpression);
+                var alkoi = start == null ? LocalDate.ofEpochDay(0) : start;
+                var loppui = getLocalDate(o, "$.päättymispäivä");
+
+                return new KoulutusDto(null, localizedNimi, localizedKuvaus, alkoi, loppui, null);
+              })
+          .toList();
```

Comment:
It looks like the usage is JSONPath is quite trivial -- perhaps one could just use Jackson JsonNode, potentially with JsonPointer and avoid the extra dependency? For PoC this is OK, but something to consider in the future.